plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/DefaultHttpResponseSender.java

```diff
@@ -50,62 +61,180 @@
-      httpResponse.addHeader(header.getKey(), header.getValue());
-    }
+    handleResponseHeaders(response, httpResponse);
 
     // add status followed by payload if we have one
     Status status = response.getStatus();
     if (status.isSuccessful() || response instanceof PayloadResponse) {
       httpResponse.setStatus(status.getCode());
       if (response instanceof PayloadResponse) {
-        Payload payload = ((PayloadResponse) response).getPayload();
-        log.trace("Attaching payload: {}", payload);
-
-        if (payload.getContentType() != null) {
-          httpResponse.setContentType(payload.getContentType());
-        }
-        if (payload.getSize() != Payload.UNKNOWN_SIZE) {
-          httpResponse.setContentLengthLong(payload.getSize());
-        }
+        final PayloadResponse payloadResponse = (PayloadResponse) response;
+        log.trace("Attaching payload: {}", payloadResponse.getPayload());
+        handlePayloadHeaders(payloadResponse.getPayload(), httpResponse);
 
-        if (payload instanceof BlobPayload) {
-          final BlobPayload decoratedPayload = (BlobPayload) payload;
-          final DateTime lastModified = decoratedPayload.getLastModified();
-          if (lastModified != null && !httpResponse.containsHeader(HttpHeaders.LAST_MODIFIED)) {
-            httpResponse.setDateHeader(
-                HttpHeaders.LAST_MODIFIED,
-                lastModified.getMillis()
-            );
-          }
-          final HashAlgorithm hashAlgorithm = Iterables.getFirst(decoratedPayload.getHashAlgorithms(), null);
-          if (hashAlgorithm != null) {
-            final HashCode hashCode = decoratedPayload.getHashCodes().get(hashAlgorithm);
-            if (hashCode != null && !httpResponse.containsHeader(HttpHeaders.ETAG)) {
-              httpResponse.setHeader(
-                  HttpHeaders.ETAG,
-                  hashCode.toString()
-              );
-            }
+        if (request != null && !HttpMethods.HEAD.equals(request.getAction())) {
+          try (InputStream input = payloadResponse.getPayload().openInputStream()) {
+            sendContent(input, httpResponse);
           }
         }
+      }
+    }
+    else {
+      httpResponse.sendError(status.getCode(), status.getMessage());
+    }
+  }
 
-        if (request != null && !HttpMethods.HEAD.equals(request.getAction())) {
-          try (InputStream input = payload.openInputStream(); OutputStream output = httpResponse.getOutputStream()) {
-            ByteStreams.copy(input, output);
+  /**
+   * Set HTTP response headers based on {@link Response}.
+   */
+  protected void handleResponseHeaders(final Response response,
+                                       final HttpServletResponse httpResponse)
+  {
+    // add response headers
+    for (Map.Entry<String, String> header : response.getHeaders()) {
+      httpResponse.addHeader(header.getKey(), header.getValue());
+    }
+  }
```

Comment:
curious why are you making protected methods for a class that is never sub-classed?